Fix validation of variable lengths string datasets in Zarr

stephprince opened the pull request:
## Motivation

Fix https://github.com/hdmf-dev/hdmf-zarr/issues/221.


Zarr does not natively support variable length strings, and validation of vlen string arrays uses the HDF5 specific `data.dtype.metadata` field. This field is None for Zarr arrays resulting in the data type not being correctly extracted during validation using the `get_type` method.

## How to test the behavior?

Validate an NWB zarr file with an io object input

```python
import datetime
import uuid

import hdmf_zarr
import numpy
import pynwb

nwbfile = pynwb.NWBFile(
    session_description="",
    identifier=str(uuid.uuid4()),
    session_start_time=datetime.datetime.now().astimezone()
)
regular_timestamps = numpy.arange(1.2, 11.2, 2)
timestamps_length = len(regular_timestamps)
time_series = pynwb.TimeSeries(
    name="test_time_series",
    data=numpy.zeros(shape=(timestamps_length, timestamps_length - 1)),
    timestamps=regular_timestamps,
    unit="",
)
nwbfile.add_acquisition(time_series)

nwbfile_path = "test_validation_time_series.nwb.zarr"
with hdmf_zarr.NWBZarrIO(path=nwbfile_path, mode="w") as io:
    io.write(nwbfile)

with hdmf_zarr.NWBZarrIO(path=nwbfile_path, mode="r") as io:
    invalidations = pynwb.validate(io=io)
    print(invalidations) # should be empty []
```

## Checklist

- [x] Did you update `CHANGELOG.md` with your changes?
- [X] Does the PR clearly describe the problem and the solution?
- [X] Have you reviewed our [Contributing Guide](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst)?
- [X] Does the PR use "Fix #XXX" notation to tell GitHub to close the relevant issue numbered XXX when the PR is merged?

stephprince committed:
add condition for vlen strs in zarr validation

stephprince committed:
add tests for validation of vlen string datasets from zarr arrays

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stephprince committed:
update CHANGELOG

stephprince committed:
fix get_type function complexity

rly commented:
Thanks @stephprince ! Since variable length string datasets could also use ASCII encoding, could you please also add a test for that case? I think it will work because `NWBFile.file_create_date` is a dataset with dtype isodatetime (which is written as vlen ascii), and that hasn't raised any issues yet, but it would be good to double check and add a test while we're in this part of the code.

Could you also please add type hints to the new functions?

rly commented:
Here's an example of a vlen bytes/ascii dataset: https://github.com/hdmf-dev/hdmf/blob/bbb465875e9dfe9cb6035a552c8a1caba6150925/tests/unit/test_io_hdf5_h5tools.py#L3843C21-L3843C108

stephprince committed:
add test for vlen bytes/ascii dataset

stephprince committed:
add type hints to new helper functions

stephprince committed:
fix line length

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rly commented:
Thanks! According to the validator code and the language docs, ASCII data is allowed for a dataset spec with utf8/text dtype, but the reverse is not true. Could you add a test case with a dataset spec with ascii/bytes dtype and check that UTF8 data fails? We might not have this tested in hdf5, so it would also be nice to add that test if it doesn't exist.

stephprince committed:
add utf for ascii tests, add hdf5 dataset tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stephprince committed:
update test comment

stephprince commented:
> Could you add a test case with a dataset spec with ascii/bytes dtype and check that UTF8 data fails? We might not have this tested in hdf5, so it would also be nice to add that test if it doesn't exist.

Updated to check that utf data fails when bytes is specified in the spec and added a test with an hdf5 dataset! Let me know if what I added was what you had in mind.

rly commented on tests/unit/validator_tests/test_validate.py:
```diff
@@ -1605,3 +1606,133 @@ def test_scalar_instead_of_array(self):
+        builder = GroupBuilder('my_foo',
+                               attributes={'data_type': 'Foo', 'attr1': 'a string attribute'},
+                               datasets=[DatasetBuilder('data', ['string1', 'string2', 'string3'])])
+        results = self.runBuilderRoundTrip(builder, expected_errors=1)
```
Rather than rely on the build system and writing a file, I think it would be simpler to create a temporary hdf5 dataset:

```python
import h5py
from hdmf.backends.hdf5.h5tools import H5_BINARY

f = h5py.File(name='temp_in_memory_file.h5', driver='core', backing_store=False)
dset = f.create_dataset('data', data=['string1', 'string2', 'string3'], dtype=H5_BINARY)
```
and use this dataset in the `DatasetBuilder`. Then the tests are nicely parallel between HDF5 and Zarr.

stephprince replied:
Ah that's much simpler - thanks! I did need to still wrap it as a `StrDataset` to mimic the behavior with HDF5IO since data read directly from the dset object would be in bytes. But I think now It still tests that UTF data will fail where ascii/bytes is specified

stephprince committed:
update hdf5 vlen string validation test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stephprince committed:
remove file setup/teardown

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
